### ISSUE TITLE
Scrollbar line-spacing display issue

### ISSUE THREAD

**Issue by jcmkk3** [minad/corfu#1]
Hi Daniel,

Thanks for releasing another delightfully focused package. I am running into an issue with the "scrollbar" display on my machine. It looks to be caused by the fact that I adjust my line-spacing from the default. It's not really the end of the world, but I figured that I'd report it.

The below screenshots are with the default configuration and only the options set in the image. I'm using the official build of Emacs 27.2 for Windows.

![image](https://user-images.githubusercontent.com/45919695/114924332-b7fe5400-9de2-11eb-8104-5677afcf8138.png)
![image](https://user-images.githubusercontent.com/45919695/114924350-be8ccb80-9de2-11eb-959f-24ec30a88e73.png)

**Comment by minad** [comment]
Hmm, that's ugly. The problem is the line spacing since I am using these box drawing characters to create a smaller scrollbar. For now I suggest to adjust the faces: Set foreground=background color for corfu-bar and corfu-border and chose appropriate colors.

In the long term it may be better to fix this differently by using svgs or bitmaps. But hopefully the overlay popup code can be improved and simplified a bit more. It is a bit of a mess with an overlay for every line.

**Comment by protesilaos** [comment]
Just to note that I also got this. It seems to depend on the font family. The Hack font gives me a dashed scrollbar, while Source Code Pro looks fine.

![Screenshot from 2021-04-15 23-03-26](https://user-images.githubusercontent.com/12828033/114931890-84054c00-9e3f-11eb-8708-5fd66b28eba6.png)
![Screenshot from 2021-04-15 23-03-43](https://user-images.githubusercontent.com/12828033/114931894-85cf0f80-9e3f-11eb-8cf8-e7e089eb419c.png)

**Comment by minad** [comment]
Yes, I will probably replace this with some dynamically generated images.

**Comment by jaor** [comment]
On Thu, Apr 15 2021, Protesilaos Stavrou wrote:

> Just to note that I also got this. It seems to depend on the font
> family. The Hack font gives me a dashed scrollbar, while Source Code
> Pro looks fine.

same thing for me, in an emacs 28.  i've found a couple annoyances with
vertical spacing in hack in other contexts though, so this might not be
a real issue in corfu (or maybe it's a shared issue).  i guess an option
to disable the scrollbar would be out of scope?

**Comment by minad** [comment]
I pushed a version which uses images for the borders. This should hopefully resolve the issues and create new more problematic issues. And yes, you can disable the scrollbar, by adjusting the faces ;)

**Comment by jaor** [comment]
On Thu, Apr 15 2021, Daniel Mendler wrote:

> I pushed a version which uses images for the borders. This should
> hopefully resolve the issues and create new more problematic
> issues.

hmm, i just updated (including the forced update), and corfu-complete is
giving me this error (no matter what completion):

      Debugger entered--Lisp error: (wrong-type-argument number-or-marker-p nil)
        -(5641 nil)
        (max 0 (- (point) beg))
        (let* ((pt (max 0 (- (point) beg))) (content (buffer-substring-no-properties beg end)) (metadata (completion-metadata (substring content 0 pt) table pred))) (let* ((val (completion-try-completion content table pred pt metadata))) (if (consp val) (let* ((x38 (car-safe val)) (x39 (cdr-safe val))) (if (let* ((newstr x38)) (not (equal content newstr))) (let ((newstr x38) (newpt x39)) (ignore newstr) (let (... ... ... ... ...) (unwind-protect ... ...))))))))
        (let ((beg x30) (end x32) (table x34) (pred x36)) (let* ((pt (max 0 (- (point) beg))) (content (buffer-substring-no-properties beg end)) (metadata (completion-metadata (substring content 0 pt) table pred))) (let* ((val (completion-try-completion content table pred pt metadata))) (if (consp val) (let* ((x38 (car-safe val)) (x39 (cdr-safe val))) (if (let* (...) (not ...)) (let (... ...) (ignore newstr) (let ... ...))))))))
        (progn (ignore (null x37)) (let ((beg x30) (end x32) (table x34) (pred x36)) (let* ((pt (max 0 (- (point) beg))) (content (buffer-substring-no-properties beg end)) (metadata (completion-metadata (substring content 0 pt) table pred))) (let* ((val (completion-try-completion content table pred pt metadata))) (if (consp val) (let* ((x38 ...) (x39 ...)) (if (let* ... ...) (let ... ... ...))))))))
        (let* ((x36 (car-safe x35)) (x37 (cdr-safe x35))) (progn (ignore (null x37)) (let ((beg x30) (end x32) (table x34) (pred x36)) (let* ((pt (max 0 (- ... beg))) (content (buffer-substring-no-properties beg end)) (metadata (completion-metadata (substring content 0 pt) table pred))) (let* ((val (completion-try-completion content table pred pt metadata))) (if (consp val) (let* (... ...) (if ... ...))))))))
        (progn (ignore (consp x35)) (let* ((x36 (car-safe x35)) (x37 (cdr-safe x35))) (progn (ignore (null x37)) (let ((beg x30) (end x32) (table x34) (pred x36)) (let* ((pt (max 0 ...)) (content (buffer-substring-no-properties beg end)) (metadata (completion-metadata ... table pred))) (let* ((val ...)) (if (consp val) (let* ... ...))))))))
        (let* ((x34 (car-safe x33)) (x35 (cdr-safe x33))) (progn (ignore (consp x35)) (let* ((x36 (car-safe x35)) (x37 (cdr-safe x35))) (progn (ignore (null x37)) (let ((beg x30) (end x32) (table x34) (pred x36)) (let* ((pt ...) (content ...) (metadata ...)) (let* (...) (if ... ...))))))))
        (progn (ignore (consp x33)) (let* ((x34 (car-safe x33)) (x35 (cdr-safe x33))) (progn (ignore (consp x35)) (let* ((x36 (car-safe x35)) (x37 (cdr-safe x35))) (progn (ignore (null x37)) (let ((beg x30) (end x32) (table x34) (pred x36)) (let* (... ... ...) (let* ... ...))))))))
        (let* ((x32 (car-safe x31)) (x33 (cdr-safe x31))) (progn (ignore (consp x33)) (let* ((x34 (car-safe x33)) (x35 (cdr-safe x33))) (progn (ignore (consp x35)) (let* ((x36 (car-safe x35)) (x37 (cdr-safe x35))) (progn (ignore (null x37)) (let (... ... ... ...) (let* ... ...))))))))
        (progn (ignore (consp x31)) (let* ((x32 (car-safe x31)) (x33 (cdr-safe x31))) (progn (ignore (consp x33)) (let* ((x34 (car-safe x33)) (x35 (cdr-safe x33))) (progn (ignore (consp x35)) (let* ((x36 ...) (x37 ...)) (progn (ignore ...) (let ... ...))))))))
        (let* ((x30 (car-safe completion-in-region--data)) (x31 (cdr-safe completion-in-region--data))) (progn (ignore (consp x31)) (let* ((x32 (car-safe x31)) (x33 (cdr-safe x31))) (progn (ignore (consp x33)) (let* ((x34 (car-safe x33)) (x35 (cdr-safe x33))) (progn (ignore (consp x35)) (let* (... ...) (progn ... ...))))))))
        (progn (ignore (consp completion-in-region--data)) (let* ((x30 (car-safe completion-in-region--data)) (x31 (cdr-safe completion-in-region--data))) (progn (ignore (consp x31)) (let* ((x32 (car-safe x31)) (x33 (cdr-safe x31))) (progn (ignore (consp x33)) (let* ((x34 ...) (x35 ...)) (progn (ignore ...) (let* ... ...))))))))
        (if (>= corfu--index 0) (corfu-insert) (progn (ignore (consp completion-in-region--data)) (let* ((x30 (car-safe completion-in-region--data)) (x31 (cdr-safe completion-in-region--data))) (progn (ignore (consp x31)) (let* ((x32 (car-safe x31)) (x33 (cdr-safe x31))) (progn (ignore (consp x33)) (let* (... ...) (progn ... ...))))))))
        corfu-complete()
        funcall-interactively(corfu-complete)
        call-interactively(corfu-complete record nil)
        command-execute(corfu-complete record)
        execute-extended-command(nil "corfu-complete" nil)
        funcall-interactively(execute-extended-command nil "corfu-complete" nil)
        call-interactively(execute-extended-command nil nil)
        command-execute(execute-extended-command)

if i try corfu-next instead, first i see an error that says:

    Error in post-command-hook (corfu--post-command-hook): (error "Not enough arguments for format string")

and then i get completions without any overlay being shown.

> And yes, you can disable the scrollbar, by adjusting the faces ;)

yeah, i figured that much :)  a minor nit/suggestion: why not calling
the face used for the background of the bar corfu-bar-background instead
of the (to me, at least) cryptic corfu-border? to me the last name
suggested the color of, well, a border around the overlay, and the
corfu-bar name suggested that its foreground would be used for the
moving part of the bar, and its background for the bar's background
(although i understand why that is not direct to implement), so i would
even consider renaming it to corfu-bar-foreground.  not very important,
but company's face names always have me scratching my head and had a
feeling of deja vu with those ones :)

**Comment by minad** [comment]
Can you please recheck that you are using the newest version or check if only the last commit makes problems? The error you are observing points to the completion-in-region--data being invalid (beg=nil).

> why not calling
the face used for the background of the bar corfu-bar-background instead
of the (to me, at least) cryptic corfu-border?

Now it is a real border. But I will think if I can improve how the faces are organized.

**Comment by jaor** [comment]
On Thu, Apr 15 2021, Daniel Mendler wrote:

> Can you please recheck that you are using the newest version or check
> if only the last commit makes problems?

a497acc42d4021a8369c41993102ddd5c9ebdee9 (which AFAICT is HEAD) gives me
the errors.  HEAD^ (e0198b4335add183e0673c2b224e3d6c3ac09cab) works as before...

**Comment by minad** [comment]
Okay, thanks!

**Comment by jaor** [comment]
On Thu, Apr 15 2021, Daniel Mendler wrote:

> Okay, thanks!

thanks to /you/.  in case that might matter, i'm on today's emacs master
(not the nativecomp branch).

**Comment by protesilaos** [comment]
> Closed #1 via 219afe2.

Hello Daniel!

I have been using Corfu for a few hours now.  It works well thus far.
It also comes at a time when my musing with the minibuffer have rendered
ineffective the Consult function for completion in region: with the
default setup the Completions' buffer pops up, but the minibuffer is not
activated, so it is not a good setup (note that I have not debugged this
yet, so there is a good chance that I have messed up something on my
end).

The only feature I would want from Corfu (or a tool along those lines)
is to somehow extend dabbrev (yes, I know the problems---just expressing
my wish for Santa).

And since I am on the wishlist, how complicated would it be to pipe
those completion candidates into a bespoke buffer?  Maybe it could be
done at the user level to keep the SLOC count in check.

Thanks again for your contributions!

-- 
Protesilaos Stavrou
protesilaos.com

**Comment by minad** [comment]
@protesilaos 

Thank you for giving it a try!

> It also comes at a time when my musing with the minibuffer have rendered
ineffective the Consult function for completion in region: with the
default setup the Completions' buffer pops up, but the minibuffer is not
activated, so it is not a good setup (note that I have not debugged this
yet, so there is a good chance that I have messed up something on my
end).

I think `consult-completion-in-region` only works well if you use minibuffer completion.

> The only feature I would want from Corfu (or a tool along those lines)
is to somehow extend dabbrev (yes, I know the problems---just expressing
my wish for Santa).

Yes, a dabbrev capf backend should be doable as a separate project as we discussed above.

> And since I am on the wishlist, how complicated would it be to pipe
those completion candidates into a bespoke buffer?  Maybe it could be
done at the user level to keep the SLOC count in check.

Maybe one could add embark support, the same applies to company.
